src/main/java/com/google/codeu/data/Datastore.java

```diff
@@ -85,6 +87,59 @@ public void storeMessage(Message message) {
     return messages;
   }
 
+  /**
+   * Gets messages sent between the logged-in user and another user.
+   *
+   * @return a list of messages sent between the logged-in user and another user, or empty list
+   * if logged-in user or other user have never sent a message to each other. List is sorted by time ascending.
+   */
+  public List<Message> getMessagesBetweenTwoUsers(String loggedInUser, String otherUser) {
+    List<Message> messages = new ArrayList<>();
+
+    // Messages between two people, where logged-in user is the sender
+    Filter messagesSentByLoggedInUser = new Query.FilterPredicate("user", FilterOperator.EQUAL, loggedInUser);
+    Filter messagesReceivedByOtherUser = new Query.FilterPredicate("recipient", FilterOperator.EQUAL, otherUser);
+
+    // All the messages sent by logged-in user to the other user
+    Filter loggedInUserMessages = CompositeFilterOperator.and(messagesSentByLoggedInUser, messagesReceivedByOtherUser);
+
+    // Messages between two people, where logged-in user is the recipient
+    Filter messagesSentByOtherUser = new Query.FilterPredicate("user", FilterOperator.EQUAL, otherUser);
+    Filter messagesReceivedByLoggedInUser = new Query.FilterPredicate("recipient", FilterOperator.EQUAL, loggedInUser);
+    
+    // All the messages recieved by logged-in user sent by the other user
+    Filter otherUserMessages = CompositeFilterOperator.and(messagesSentByOtherUser, messagesReceivedByLoggedInUser);
+
+    // All the messages between the two users
+    Filter directMessages = CompositeFilterOperator.or(loggedInUserMessages, otherUserMessages);
+
+    Query query =
+        new Query("Message")
+            .setFilter(directMessages)
+            .addSort("timestamp", SortDirection.ASCENDING);
+    PreparedQuery results = datastore.prepare(query);
+
+    for (Entity entity : results.asIterable()) {
+      try {
```

Comment:
Which of these statements to do you expect to fail?

In general, a try statement should be as narrowly scoped as possible. Consider using helper functions that do a single operation within a "try" to keep the scope to a minimum. For example, you might have something like:

       private Optional<String> getStringProperty(PropertyContainer container, String propertyName) {
         if (!container.hasProperty(propertyName)) {
           return Optional.<string>absent();
         }

         String value = null;
         try {
           value = (String) container.getProperty(propertyName);
         } catch (ClassCastException wrongType) {
            logging.atSevere().withCause(wrongType).log("Property \"" + propertyName + "\" exists but is not a String.");
         }
         return Optional.<String>ofNullable(value);
     }

... assuming that this is the aspect of the code below you expect might fail.

Reply:
What do you mean by keeping the try scope to minimum? Only check a small number of things within a try? So if I expected all the things within the try statement, as it is right now, to fail, I should have a separate try catch for every property?

Reply:
Yep, that's exactly right; the number of operations performed inside a "try { ... }" should be the minimum number of statements possible. To keep this code from being ugly with lots of separate "try { ... }" blocks, I would use a subroutine (or multiple subroutines) such as the "getStringProperty" function that I prototyped above, where the subroutine provides for a path to graceful error handling.

For example, with the "getStringProperty" code that I suggested above, you could do something like:

       String text = getStringProperty(entity, "text").orElse("");

... which causes an empty string to be used as a fallback in the case of an error.

Reply:
What is Optional.<String>absent()?  I don't see that in the Java documentation

Reply:
Sorry for the mistake/confusion. The `Optional` type was first introduced in Guava (Google's Java core libraries) before it became standardized as part of Java 8; when it was incorporated in Java 8, there were some minor naming changes to some of the methods.

I meant to say `Optional.<String>empty`:
https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#empty--

`Optional.absent` is the legacy name in Guava that isn't available in Java 8:
https://google.github.io/guava/releases/19.0/api/docs/com/google/common/base/Optional.html#absent()

Reply:
It seems like I can't use the FluentLogger in the code

Reply:
You will need to update the `pom.xml` file to add a dependency on Flogger and add the relevant import statements for it to work.

The repository page has the instructions:
https://github.com/google/flogger

You can see the available `<dependencies>` that can be added in the `pom.xml` file here:
https://mvnrepository.com/artifact/com.google.flogger

If you click through the links to a specific version (you need `flogger` and `flogger-system-backend`), it will give you a `<dependency>...</dependency>` XML snippet that you can copy+paste into the `pom.xml` file.